panoramapublic/src/org/labkey/panoramapublic/chromlib/ChromLibStateImporter.java

```diff
@@ -0,0 +1,406 @@
+                dbMoleculePrecursors.forEach(p -> _precursorKeyMap.put(p.getKey(), p.getId()));
+            }
+
+            LibGeneralPrecursor precursor = parsePrecursor(row, _currentPepGrp);
+            if (RepresentativeDataState.NotRepresentative != precursor.getRepresentativeDataState())
+            {
+                List<Long> precursorIdsForState = _precursorsForState.computeIfAbsent(precursor.getRepresentativeDataState(), l -> new ArrayList<>());
+                precursorIdsForState.add(precursor.getId());
+            }
+        }
+
+        private void updatePrecursorState(List<Long> precursorIds, RepresentativeDataState state, Container container, User user, TargetedMSService svc, Logger log)
+        {
+            updateState(svc.getTableInfoGeneralPrecursor(), precursorIds, state, container, user, svc, log);
+        }
+
+        @Override
+        void onNextRun()
+        {
+            super.onNextRun();
+            _currentPepGrp = null;
+            _precursorsForState = new HashMap<>();
+        }
+
+        @Override
+        void updateRows()
+        {
+            for (RepresentativeDataState state: _precursorsForState.keySet())
+            {
+                updatePrecursorState(_precursorsForState.get(state), state, _container, _user, _svc, _log);
+            }
+            _precursorsForState.clear();
+        }
+
+        private LibGeneralPrecursor parsePrecursor(Map<String, Object> row, LibPeptideGroup peptideGroup)
+        {
+            double precursorMz = (Double) row.get(PREC_MZ);
+            int precursorCharge = (Integer) row.get(PREC_CHARGE);
+            String precursorState = (String) row.get(PREC_STATE);
+            String modifiedSeq = (String) row.get(PREC_MOD_SEQ);
+
+            LibGeneralPrecursor precursor;
+            if (!StringUtils.isBlank(modifiedSeq))
+            {
+                precursor = new LibPrecursor(peptideGroup.getId(), precursorMz, precursorCharge, RepresentativeDataState.valueOf(precursorState), modifiedSeq);
+            }
+            else
+            {
+                String customIonName = (String) row.get(PREC_MOL_CUSTOM_ION_NAME);
+                String ionFormula = (String) row.get(PREC_MOL_ION_FORMULA);
+                Double massMonoIsotopic = (Double) row.get(PREC_MOL_MASS_MONOISOTOPIC);
+                Double massAverage = (Double) row.get(PREC_MOL_MASS_AVERAGE);
+                precursor = new LibMoleculePrecursor(peptideGroup.getId(), precursorMz, precursorCharge, RepresentativeDataState.valueOf(precursorState),
+                        customIonName, ionFormula, massMonoIsotopic, massAverage);
+            }
+
+            Long generalPrecursorId = _precursorKeyMap.get(precursor.getKey());
+            if (generalPrecursorId == null)
+            {
+                throw new IllegalStateException(String.format("Expected a row for precursor %s in the peptide group %s. Skyline document '%s'. Container '%s'.",
```

Comment:
Optional: I generally recommend using "folder" instead of "container" in user-facing error messages. And passing in _container.getPath() to avoid the rest of the toString() content.